scapy/fields.py

```diff
@@ -774,50 +774,72 @@ def i2repr(self, pkt, x):
 
 class _EnumField(Field):
     def __init__(self, name, default, enum, fmt = "H"):
-        i2s = self.i2s = {}
-        s2i = self.s2i = {}
-        if type(enum) is list:
-            keys = range(len(enum))
+        if isinstance(enum, tuple):
+            assert(len(enum) == 2 and callable(enum[0]) and callable(enum[1]))
+            self.i2s_cb = enum[0]
+            self.s2i_cb = enum[1]
+            self.i2s = None
+            self.s2i = None
         else:
-            keys = enum.keys()
-        if any(type(x) is str for x in keys):
-            i2s, s2i = s2i, i2s
-        for k in keys:
-            i2s[k] = enum[k]
-            s2i[enum[k]] = k
+            i2s = self.i2s = {}
+            s2i = self.s2i = {}
+            self.i2s_cb = None
+            self.s2i_cb = None
+            if type(enum) is list:
+                keys = range(len(enum))
+            else:
+                keys = enum.keys()
+            if any(type(x) is str for x in keys):
+                i2s, s2i = s2i, i2s
+            for k in keys:
+                i2s[k] = enum[k]
+                s2i[enum[k]] = k
         Field.__init__(self, name, default, fmt)
+
     def any2i_one(self, pkt, x):
         if type(x) is str:
-            x = self.s2i[x]
+            if self.s2i is None:
+                x = self.s2i_cb(x)
+            else:
+                x = self.s2i[x]
         return x
+
     def i2repr_one(self, pkt, x):
-        if self not in conf.noenum and not isinstance(x,VolatileValue) and x in self.i2s:
-            return self.i2s[x]
+        if self not in conf.noenum and not isinstance(x,VolatileValue):
+            if self.i2s is None:
```

Comment:
I'd prefer:
```Python
try:
    return self.i2s[x]
except KeyError:
    pass
except TypeError:
    return self.i2s_cb(x)
```

By the way, maybe we could imagine that `.i2s_cb` returns `None` when it has no specific name for a value (that would be the equivalent of a missing key in `.i2s`. In that case, the code could be adapted as:

```Python
try:
    return self.i2s[x]
except KeyError:
    pass
except TypeError:
    val = self.i2s_cb(x)
    if val is not None:
        return val
```